File: src/main/java/Verifier.java
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Verifier {

    private final List<Integer> nodes;

    Verifier(int... path) {
        nodes = IntStream.of(path).boxed().collect(Collectors.toList());
    }

    boolean isHamiltonianPath(int n) {
        boolean allNodes = IntStream.rangeClosed(1, nodes.size()).allMatch(nodes::contains);
        boolean allSquares = IntStream.range(0, nodes.size() -1).allMatch(v -> isSquare(nodes.get(v) + nodes.get(v + 1)));
        return nodes.size() == n && allNodes && allSquares;
    }

    boolean isHamiltonianCycle(int n) {
        int number = nodes.size();
        return isHamiltonianPath(n) && isSquare(nodes.get(0) + nodes.get(number - 1));
    }

    private boolean isSquare(int number) {
        double sqrt = Math.sqrt(number);
        return sqrt == Math.ceil(sqrt);
    }
}

File: src/main/java/util/Verifier.kt
package util

class Verifier(private val nodes: List<Int>, private val squareCache: SquareCache) {

    fun isHamiltonianPath(length: Int): Boolean {
        if (nodes.size != length) return false
        val allNodes = containsEachNodeOnce()
        val allSquares = (0 until nodes.size - 1).all { v -> squareCache.contains(nodes[v] + nodes[v + 1]) }
        return allNodes && allSquares
    }

    private fun containsEachNodeOnce(): Boolean {
        val seen = IntArray(nodes.size)
        for (node in nodes) {
            val nodeIdx = node - 1
            seen[nodeIdx] += 1
            if (seen[nodeIdx] > 1) return false
        }
        return true
    }

    fun isHamiltonianCycle(length: Int): Boolean =
            isHamiltonianPath(length) && squareCache.contains(nodes[0] + nodes[nodes.size - 1])

}
